testinfra/backend/vagrant.py

```diff
@@ -0,0 +1,333 @@
+        if self._last_refreshed is None:
+            self._last_refreshed = time.time() - self.status_refresh_interval
+        time_between_last_run = round(time.time() - self._last_refreshed, 2)
+
+        should_refresh = time_between_last_run >= self.status_refresh_interval
+        return should_refresh
+
+    def _stop(self, destroy=False):
+        run_args = ['vagrant']
+        if destroy:
+            run_args.append('destroy')
+        else:
+            run_args.append('halt')
+        run_args.append('--force')
+        run_args.append(self.box)
+        run_args = ' '.join(run_args)
+        out = self.run_vagrant(run_args)
+        return out
+
+    def _validate_requirements(self):
+        if not self.has_vagrant and self.setting.get('VALIDATE_HAS_VAGRANT', True):
+            raise RuntimeError('Vagrant must be installed in order ' +
+                               'to use this fixture')
+
+        if not self.has_vagrantfile:
+            raise RuntimeError(
+                'Unable to find Vagrantfile "{}" and I am in directory {}'.format(
+                    self.vagrantfile,
+                    os.getcwd()
+                )
+            )
+
+    def _refresh_status(self):
+        out = self.run_vagrant('vagrant status %s', self.box)
+        status = re.sub(r'  +| \(', '|', out).rstrip(')').split('|')[1]
+
+        # update the counter that controls this cache
+        self._last_refreshed = time.time()
+
+        if status in ('aborted', 'not created', 'paused', 'poweroff', 'saved',
+                      'suspended', 'not running'):
+            as_running = False
+            as_not_running = True
+            created = status != 'not created'
+        elif status == 'running':
+            as_running = True
+            as_not_running = False
+            created = True
+        else:
+            raise NotImplementedError('This is a bug, un-handled status from ' +
+                                      '`vagrant status {}` '.format(self.box) +
+                                      'got "{}"'.format(status)
+                                      )
+
+        def cmd():
+            def add(key, val):
+                setattr(cmd, key, val)
+            cmd.add = add
+            return cmd
+
+        cmd = cmd()
+        cmd.add('to_be_running', as_running)
+        cmd.add('not_to_be_running', as_not_running)
+        cmd.add('is_running', as_running)
+        cmd.add('is_not_running', as_not_running)
+        cmd.add('running', as_running)
+        cmd.add('not_running', as_not_running)
+        cmd.add('is_created', created)
+        cmd.add('is_not_created', not created)
+
+        self._status = cmd
+        return self._status
+
+    def _refresh_ssh_config(self):
+        if self._ssh_config is None:
+            self._ssh_config = self.run_vagrant('vagrant ssh-config %s',
+                                                self.box
+                                                )
```

Comment:
My original intention with this was for testinfra to do the setup:
1. Start the Vagrant box
2. Prep the vagrant box with needed libraries so that our tests will run
3. Provide way to simplify creating the hooks that could be used during your CI testing and using Vagrant through testinfra where you focus on the outcome of tests rather than fussing around with getting vagrant to work. 
4. Testing shell scripts is hard, testing python is much easier and this library is tested and it should work with the latest versions of Vagrant and I totally plan on helping maintain it if you find this library helpful for testinfra.
5. This backend was primarily created to be used as a test fixture for (spin up, run your tests, tear it down) and repeat and only focusing on solid building blocks.

You won't hurt my feelings if you say you don't think it will be a good fit for this project.  Either way, I'm going to keep this code around in my repo incase others find it useful.